Send updated offers to subscribers after publisher renegotiations

danxuliu opened the pull request:
Although related, this is independent from #191.

Tested also when running against a proxy :-)

When a publisher has a connection the publisher can update the connection (for example, to add a video track to an audio only connection) by sending an updated offer to Janus. Janus detects that, adjusts the connection and then sends back an answer. Once the publisher connection is updated Janus starts a renegotiation for the subscribers and generates the offers to be sent to them.

The signaling server did not handle the event, so the offers were not sent and the subscriber connections were not updated. Now the offers are sent as needed, which makes possible for the renegotiation to be completed by the clients.

In this case the `offer` message will also include an `update` parameter so clients can differentiate between offers to create new connections or update the existing one.

A new feature flag, `publisher-update`, is also introduced to notify clients whether updating publisher connections is supported or not (supported as in _it will cause renegotiation offers to be sent for the subscribers_). This is needed for the clients to differentiate if they can just send an updated offer or if they need to force a reconnection. However, would it be better to drop the flag here and add a single flag about renegotiations once #191 is ready?

Some things I am not very sure about (I am quite unsure about any change made by me to this code, but specially about these :-P ):
- `OnOffer` should be called `OnUpdatedSubscriberOffer`, `OnConfiguredSubscriber` or something like that? `OnOffer` may be a bit misleading, as it looks like it would be called for any offer, and not only when a negotiation is needed in Janus for a subscriber and thus the offer needs to be sent to the other peer.
- Second parameter of `OnOffer` is defined as `map[string]interface{}`, as it carries the offer provided by Janus in `event.Jsep` (_mcu_janus.go_), which (currently) is a string indexed map with fields `sdp` and `type`. However, in _mcu_proxy.go_ `msg.Payload["offer"]` is `interface{}`, so the compiler complains that it needs to be type asserted to `map[string]interface{}`. Is there a better way to do it? Note that if the second parameter of `OnOffer` is changed to `interface{}` then the type assertion will be needed in _clientsession.go_ in `Payload:offer`.
- Could there be any possible race condition if a new subscriber joins when the publisher is updated?

fancycode commented on api_signaling.go:
```diff
@@ -325,6 +325,7 @@ const (
 	// Features for all clients.
 	ServerFeatureMcu                   = "mcu"
 	ServerFeatureSimulcast             = "simulcast"
+	ServerFeaturePublisherUpdate      = "publisher-update"
```
```suggestion
	ServerFeaturePublisherUpdate       = "publisher-update"
```

This should have been caught by the CI linter.

fancycode replied:
Linter is updated in #196

fancycode commented on mcu_janus.go:
```diff
@@ -1063,7 +1063,11 @@ func (p *mcuJanusSubscriber) handleEvent(event *janus.EventMsg) {
-			// Ignore events like selected substream / temporal layer.
+			// Handle renegotiations, but ignore other events like selected
+			// substream / temporal layer.
+			if (getPluginStringValue(event.Plugindata, pluginVideoRoom, "configured") == "ok") {
```
```suggestion
			if getPluginStringValue(event.Plugindata, pluginVideoRoom, "configured") == "ok" {
```

This should have been caught by the CI linter.

fancycode replied:
Linter is updated in #196

danxuliu replied:
_Somebody_ is too used to C-like syntax... ;-) Fixed.

fancycode reviewed (commented):
Thanks! Some initial feedback on the code. Didn't test yet (see https://github.com/nextcloud/spreed/pull/6896#issuecomment-1036009033).

> OnOffer should be called OnUpdatedSubscriberOffer, OnConfiguredSubscriber or something like that?

What about `OnUpdateOffer`?

> Second parameter of OnOffer is defined as map[string]interface{}, [...]

I updated the code to match what is done for candidates. Didn't test yet if the format for the clients is the same though.

> Could there be any possible race condition if a new subscriber joins when the publisher is updated?

That should IMHO be handled by Janus (which is the source of truth for SDPs).

fancycode commented:
Also I would combine this with #191 and introduce a single feature flag like `renegotiate` or `update-sdp`.

danxuliu commented:
> > OnOffer should be called OnUpdatedSubscriberOffer, OnConfiguredSubscriber or something like that?
>
> What about OnUpdateOffer?

It is your code, so I will rename it to whatever you tell me ;-) I have added a fixup for that.

> > Second parameter of OnOffer is defined as map[string]interface{}, [...]
>
> I updated the code to match what is done for candidates. Didn't test yet if the format for the clients is the same though.

That is the reason why I initially used `map[string]interface{}`. The [payload of the "normal" offer directly contains](https://github.com/strukturag/nextcloud-spreed-signaling/blob/76ad24ca62b188a6ac3575790a92e499bc5b2fa0/hub.go#L1823) the `sdp` and `type` [returned by Janus in the Jsep](https://github.com/strukturag/nextcloud-spreed-signaling/blob/614c5f6e21ac59775b063454341f05033d259284/mcu_janus.go#L533), it does not have an `offer` sublevel. Therefore, using
```
		Payload:  map[string]interface{}{
			"offer": offer,
		},
```
would make the payload to be different depending on the offer type. Due to this I did not add a fixup for this yet.

> > Could there be any possible race condition if a new subscriber joins when the publisher is updated?
>
> That should IMHO be handled by Janus (which is the source of truth for SDPs).

Yes, I would expect Janus to properly handle that (let's hope it does :-P ). I meant race condition in how the signaling server handles the events from Janus, as I just copy-pasted and adjusted existing code without a deep understanding ;-) Due to this I do not know if it could happen for example that the signaling server takes too long to create a subscriber object, Janus emits the configured event and then the updated offer is lost because although the subscriber exists in Janus it did not exist yet in the signaling server.

> Also I would combine this with #191 and introduce a single feature flag like `renegotiate` or `update-sdp`.

:+1: Should everything be combined in a single pull request? Or just work independently in the features and then add the final flag once everything is ready?

danxuliu commented on api_signaling.go:
```diff
@@ -647,4 +648,5 @@ type AnswerOfferMessage struct {
 	Type     string                 `json:"type"`
 	RoomType string                 `json:"roomType"`
 	Payload  map[string]interface{} `json:"payload"`
+	Update   bool                   `json:"update,omitempty"`
```
If an RTCPeerConnection handles an offer that is not an update it will end failing, but depending on the implementation it can take up to 30 seconds to fail (and in some cases I think that I saw it getting stuck, but I am not 100% sure).

In general this should not happen, but as offers are periodically requested until a connection is established in some corner cases with delayed messages and flaky connections it could happen that an offer was requested and just before receiving it a new request was made. The original connection is established, but Janus closes it due to receiving the new offer request. However, in some cases the browser could keep the connection open, so when the new offer is received it would be handled by the existing connection, rather than immediately establishing a new one.

Due to this Talk's WebUI closes the previous connection and starts a new one when an offer for the same participant is received, and this is why it needs to differentiate between normal offers, which closes the connection and start a new one, and updated offers, which apply to the existing one.

fancycode commented on mcu_janus.go:
```diff
@@ -1063,7 +1063,11 @@ func (p *mcuJanusSubscriber) handleEvent(event *janus.EventMsg) {
-			// Ignore events like selected substream / temporal layer.
+			// Handle renegotiations, but ignore other events like selected
+			// substream / temporal layer.
+			if getPluginStringValue(event.Plugindata, pluginVideoRoom, "configured") == "ok" {
```
This should check `event.Jsep` to make sure an offer has been received (`"type" = "offer" and "sdp" = "..."`) in case there are other events that can be sent by Janus. We can defer this to a follow-up PR.

danxuliu replied:
I have added a fixup for that. I tested it and as far as I can tell it works, but please check the syntax :-) Thanks!

fancycode commented on clientsession.go:
```diff
@@ -563,6 +563,34 @@ func (s *ClientSession) SetClient(client *Client) *Client {
 	return prev
 }
 
+func (s *ClientSession) sendOffer(client McuClient, sender string, streamType string, offer map[string]interface{}) {
```
Note to self: de-duplicate with https://github.com/danxuliu/nextcloud-spreed-signaling/blob/88636889d06639c8ca7810fd06e201b823ba9b97/hub.go#L1822 once this has landed.

danxuliu commented:
> The signaling server will first create the internal subscriber object and then send the subscription request to Janus, so there should be no race condition here.

Perfect!

> I don't care if you want to keep it in two PRs or want to combine it - do what fits your workflow best.

Then I would prefer to keep it as two separate PRs. Nevertheless, feel free to merge this one once it is ready or wait until #191 is ready too, I have no preference on that :-)

fancycode commented on api_signaling.go:
```diff
@@ -325,6 +325,7 @@ const (
 	// Features for all clients.
 	ServerFeatureMcu                   = "mcu"
 	ServerFeatureSimulcast             = "simulcast"
+	ServerFeaturePublisherUpdate       = "publisher-update"
```
```suggestion
	ServerFeatureUpdateSdp             = "update-sdp"
```

fancycode reviewed (commented):
Looks good now, thanks!

Could you please change the feature id to be more generic in preparation for combining this with #191 and then squash your commits?

danxuliu commented:
> Could you please change the feature id to be more generic in preparation for combining this with #191 and then squash your commits?

Done. I squashed the fixups in their respective commits, but I was not sure if you wanted all the commits squased in a single one or not, so for now I kept the commit to add the feature and the commit to add the common flag separated. If you want me to squash the commit that adds the flag into the first commit too let me know :-)

Indepedently of that, I can also rebase on current master to use the latest CI changes. Again, just let me know ;-)

fancycode commented:
Perfect, thanks!
